airbyte-cdk/python/airbyte_cdk/sources/declarative/states/dict_state.py

```diff
@@ -0,0 +1,75 @@
+#
+# Copyright (c) 2022 Airbyte, Inc., all rights reserved.
+#
+
+from enum import Enum
+from typing import Mapping, Union
+
+from airbyte_cdk.sources.declarative.interpolation.jinja import JinjaInterpolation
+from airbyte_cdk.sources.declarative.states.state import State
+
+
+def _get_max(*, name, val, other_state):
+    other_val = other_state.get(name)
+    if other_val:
+        return max(val, other_val)
+    else:
+        return val
+
+
+class StateType(Enum):
+    STR = str
+    INT = int
+
+
+class DictState(State):
+    stream_state_field = "stream_state"
+
+    def __init__(self, d: Mapping[str, str] = None, state_type: Union[str, StateType, type] = "STR", config=None):
+        if d is None:
+            d = dict()
+        if config is None:
+            config = dict()
+        self._d = d
+        if type(state_type) == str:
+            self._state_type = StateType[state_type].value
+        elif type(state_type) == StateType:
+            self._state_type = state_type.value
+        elif type(state_type) == type:
+            self._state_type = state_type
+        else:
+            raise Exception(f"Unexpected type for state_type. Got {state_type}")
+        self._interpolator = JinjaInterpolation()
+        self._context = dict()
+        self._config = config
+
+    def update_state(self, **kwargs):
+        stream_state = kwargs.get(self.stream_state_field)
+        prev_stream_state = self.get_stream_state() or stream_state
+        self._context.update(**kwargs)
+
+        self._context[self.stream_state_field] = self._compute_state(prev_stream_state)
+
+    def get_context(self):
```

Comment:
I'm not a huge fan of exposing the `context` publicly when it feels like accessing and modifying state can be done through `get_state()` and `update_state()`. Unless we had a use case for it. If we do get rid of it, we'll have to swap its use in the offset paginator.